File: Pod/ViewObject/MSAgentViewObject.m
//
// Created by P.I.akura on 2013/08/18.
// Copyright (c) 2013 P.I.akura. All rights reserved.
//
// To change the template use AppCode | Preferences | File Templates.
//


#import "MSAgentViewObject.h"
#import "TableViewAgentCategory.h"

#ifdef DEBUG
#define precondition(code) ({code});
#else
#define precondition(code) ;
#endif

@implementation MSAgentViewObject

- (id)initWithArray:(NSMutableArray *)array {
    precondition(
            assert(![_array isKindOfClass:NSMutableArray.class]);
            for (int i = 0, _len = _array.count; i < _len; i++) {
                assert(![_array[i] isKindOfClass:NSMutableArray.class]);
            });

    self = [super init];
    if (self) {
        _array = array;
    }
    return self;
}

- (void)addObject:(id)object inSection:(NSInteger)section {
    if (_array.count == 0) {
        [_array addObject:[NSMutableArray array]];
    }
    [_array[section] addObject:object];
    [_agent insertCell:[NSIndexPath indexPathForItem:[_array[section] count] - 1 inSection:section]];
}

- (void)changeObject:(id)object {
    [_agent changeUpdateCell:[self indexPathForObject:object]];
}

- (NSIndexPath *)indexPathForObject:(id)object {
    for (NSInteger i = 0, _len = _array.count; i < _len; i++) {
        for (NSInteger j = 0, _len = [_array[i] count]; j < _len; j++) {
            if ([_array[i][j] isEqual:object]) {
                return [NSIndexPath indexPathForItem:j inSection:i];
            }
        }
    }
    return nil;
}

- (void)dealloc {
    _array = nil;
}

- (NSUInteger)sectionCount {
    return _array.count;
}

- (NSUInteger)countInSection:(NSUInteger)section {
    return [_array[section] count];
}

- (id)objectAtIndexPath:(NSIndexPath *)indexPath {
    id object = _array[indexPath.section][indexPath.row];
    return self.convert ? self.convert(object) : object;
}

- (void)removeObjectAtIndexPath:(NSIndexPath *)indexPath {
    NSMutableArray *a = _array[indexPath.section];
    [a removeObjectAtIndex:indexPath.row];
    if (a.count == 0) {
        [_array removeObjectAtIndex:indexPath.section];
    }
    [_agent deleteCell:indexPath];
}

- (BOOL)existObject:(NSIndexPath *)indexPath {
    return indexPath.section < _array.count && indexPath.row < [_array[indexPath.section] count];
}

- (NSArray *)sectionObjects:(NSInteger)section {
    return _array[section];
}

@end

File: lib/ViewObject/MSAgentViewObject.swift
//
//  MSAgentViewObject.swift
//  TableViewAgent
//
//  Created by akuraru on 2014/07/30.
//  Copyright (c) 2014 P.I.akura. All rights reserved.
//

import Foundation

class MSAgentViewObject<T : NSObject, U: NSObject> : AgentViewObject<T, U> {
    weak var agent :TableViewAgent<T, U>!
    var array : [[T]]
    
    init(array :[[T]], agent: TableViewAgent<T, U>, transform: T -> U = ({t in t} as T->U)) {
        self.agent = agent
        self.array = array
        super.init(transform)
    }
    override func addObject(object :T,inSection section :Int) {
        if array.count <= section {
            array.append(Array<T>())
        }
        array[section].append(object)
        agent.insertCell(NSIndexPath(forRow: array[section].count - 1, inSection: section))
    }
    override func changeObject(object :T) {
        agent.changeUpdateCell(indexPathForObject(object))
    }
    func indexPathForObject(object :T) -> NSIndexPath! {
        for var i = 0, _len = array.count; i < _len; i++ {
            for var j = 0, _len = array[i].count; j < _len; j++ {
                if array[i][j].isEqual(object) {
                    return NSIndexPath(forRow: j, inSection: i)
                }
            }
        }
        return nil;
    }
    override func sectionCount() -> Int {
        return array.count
    }
    override func countInSection(section :Int) -> Int {
        return array[section].count
    }
    override func objectAtIndexPath(indexPath :NSIndexPath) -> U {
        return transform(array[indexPath.section][indexPath.row])
    }
    override func removeObjectAtIndexPath(indexPath :NSIndexPath) {
        var a = array[indexPath.section];
        a.removeAtIndex(indexPath.row)
        array[indexPath.section] = a
        if a.count == 0 {
            array.removeAtIndex(indexPath.section)
        }
        agent.deleteCell(indexPath)
    }
    override func existObject(indexPath :NSIndexPath) -> Bool {
        return indexPath.section < array.count && indexPath.row < array[indexPath.section].count;
    }
    override func sectionObjects(section :Int) -> U {
        return transform(array[section][0])
    }
}